zio-http/src/main/scala/zio/http/endpoint/Routes.scala

```diff
@@ -33,22 +33,19 @@ sealed trait Routes[-R, +E, M <: EndpointMiddleware] { self =>
   def toApp[R1 <: R, S](mh: RoutesMiddleware[R1, S, M])(implicit trace: Trace): App[R1] = {
     import zio.http.endpoint.internal._
 
-    val handlerTree     = HandlerTree.fromService(self)
+    val routingTree     = RoutingTree.fromRoutes(self)
     val requestHandlers = Memoized[Routes.Single[R, _ <: E, _, _, M], EndpointServer[R, _ <: E, _, _, M]] {
       handledApi =>
         EndpointServer(handledApi)
     }
 
     Http
-      .collectZIO[Request] { case (request: Request) =>
-        val handler = handlerTree.lookup(request)
+      .collectZIO[Request] {
```

Comment:
You could use `fromOptionalHandlerZIO` and then no need to evaluate the lookup function twice

Reply:
Thanks, exactly what I was looking for!